### DISCUSSION TITLE
What's the best way to get progress updates from rust into my frontend

### DISCUSSION THREAD

**Discussion by thejawker** [tauri-apps/tauri#4911] [Q&A] [answered]
I'm am working on an idea where I'd like to share progress for a few situations:
- file processing, eg: file compression, bundling, moving etc
- file uploads
- audio playback (from within rust, reason is I need to support a large set of filetypes and compressing to mp3 is not an option)

It's important that the the progress is entity scoped, eg there might be a list of uploads/audio files and they should individually update.

What is the preferred way to get this type of progress data back to the client? 

My first thought is the following:
1. client calls rust which returns a unique progress event name
2. client subscribes to this event name
3. rust keeps updating this periodically
4. client responds to the event and updates progress bar (or whatever it needs to update)
5. when done both do some cleanup

I'm thinking this should work, but I feel that there might be a better/cleaner/easier/more tauri type of approach, hence this question.

So far really digging tauri, thanks!

**Comment by abisar** [comment, marked as answer]
It's in the docs and its called [Events](https://tauri.app/v1/guides/features/events/). Here is another simple example of it:

In your JavaScript:
```javascript
import {listen} from '@tauri-apps/api/event';
import {invoke} from '@tauri-apps/api/tauri';

// Set it to a variable if you want to unlisten later
listen('stuff-happened', (event) => {
  console.log(event);
});
invoke('do_stuff');
```

And then emit the event as much as you want in rust:
```rust
// Don't forget the required includes/definitions

#[tauri::command(async)]
fn do_stuff(window: Window) {
    let mut progress: u8 = 0;

    loop {
        window.emit("stuff-happened", Payload {message: progress.to_string()}).unwrap();
        if progress > 20 {break;};
        progress += 1;
    }
}
```
`invoke` is optional. As long as you have access to `window`, you can [emit](https://docs.rs/tauri/latest/tauri/window/struct.Window.html#method.emit) events to js. The above example emits all 21 events almost at the same time, but the idea remains the same in whatever async progress you're gonna do.

> It's important that the the progress is entity scoped

You can separate them by id, event name and even window.

**Reply by thejawker** [reply]
Thank you so much @abisar for this example! This is kind of where my head was at as well. I think this should solve it.

It might be a useful package if abstracted right.
